### ISSUE TITLE
Fullscreen games break the sRGB clamp after alt-tabbing or closing the game

### ISSUE THREAD

**Issue by DerAmpelmann** [ledoge/novideo_srgb#26]
When there is an sRGB clamp applied by this program and you launch a fullscreen game (true exclusive fullscreen, not borderless fullscreen) and then alt-tab out of it or close the game again, the sRGB clamp stops being applied. When launching novideo_srgb again, the checkbox is still ticked but it is clearly not active and has to re-toggled twice to make it work again.

This is reproducible with most if not all games that offer an exclusive fullscreen mode. I experienced this with Overwatch, League of Legends and Apex Legends.

**Comment by PrismaticSpirit** [comment]
Can confirm, this also seems to occur when waking up the computer from a screensaver and when turning the monitor on and off.

**Comment by Tallulah88** [comment]
I do NOT have that issue. I just tried with Insurgency: Sandstorm and Elden Ring. I'm using fullscreen (exclusive) mode. I can't test the games that you quoted.

I'm using the latest version of novideo_srgb (3.0), Windows 11 Version 21H2 (OS build 22000.613) and a laptopt with a RTX 2060 (NVIDIA driver version 511.79). These are my novideo_srgb settings: [https://i.postimg.cc/2StDFVGV/settings.png](url)

**Comment by DerAmpelmann** [comment]
Apparently it's not an issue in all fullscreen games. I've tried some more games and can add CS:GO to the list of games that break the sRGB clamp while Cyberpunk 2077, Insurgency (don't have Sandstorm) and The Talos Principle do not break it. It seems to be an issue in esports games mostly. I assume it might have something to do with the way the exclusive fullscreen mode is implemented?

I have a secondary monitor connected and tested it with and without the DisplayCal calibration loader, I also made sure to remove the ICC profile within windows when using it inside of novideo_srgb. The settings from your screenshot are the same as I'm using myself, except I don't use dithering and chose my own ICC profile instead. The issue also happens without any ICC profile involved, when using the EDID primaries instead.

I'm on Windows 11 (21H2, build 22000.613), the NVIDIA driver version is 512.15 (on an RTX 3080).

**Comment by Tallulah88** [comment]
Then it sounds likely to be related to "Nvidia Reflex", which is present in many competitive games.

There's more info about this technology here:
[https://www.nvidia.com/en-us/geforce/news/reflex-low-latency-platform/](https://www.nvidia.com/en-us/geforce/news/reflex-low-latency-platform/)

**Comment by DerAmpelmann** [comment]
Nah, LoL and CS:GO don't even support Reflex. I tried OW without Reflex enabled a minute ago after reading your comment and this did not fix the issue either unfortunately.

I am pretty sure it is either an error in the novideo_srgb application or in the NVIDIA API. 

And looking at the second comment in this comment section it doesn't even seem to only affect games. I hope the actual issue can be found and fixed, the program is really nice besides this annoying bug and has huge potential.

**Comment by ledoge** [comment]
I can't reproduce this, the clamp still works perfectly fine after tabbing out from fullscreen games. However, I haven't updated my drivers in months, and someone on reddit reported 3 months ago that they started having issues like this with the (at the time) newest driver update. So, maybe try downgrading the driver to a version from early 2022 and see if that fixes it?

Either way, I'm pretty sure this is a driver bug and there's nothing I can do in novideo_srgb to fix (or work around) it. Though I guess technically it would be possible to have the program run in the background and listen for events like tabbing in/out of games and monitors going to sleep? Not sure how reliably that could work though... Or maybe you can try reporting this bug to NVIDIA :P

**Comment by Tallulah88** [comment]
Maybe if there's many people with this bug it can be interesting to have an option to make the program "persistent" so it reloads in the background the clamp to the GPU every X seconds for those who have this kind of issues.

**Comment by FFBBSGuT** [comment]
Having the same issue with a GTX 1660 Super and Nvidia driver version 471.11 on Win10 Pro x64 in Apex Legends and the new Overwatch 2 Beta, did not test any other games though. If it only happens in certain fullscreen games and not all of them, it might be related to the DirectX version or different implementations of this fullscreen mode (engine dependency?).

Having the program run in the background and listening for alt-tabbing or leaving fullscreen games doesn't sound like a bad idea I guess if it works reliably? This bug is the only major issue I have with novideo_sRGB and unfortunately it prevents me from using it, it's just too annoying having to retoggle the clamp every time after alt-tabbing or leaving Apex Legends. Oversaturated colors are annoying as well though so I would love to find a working solution for this problem.

I also like Tallulah88's idea with the constant reloading every X seconds, I'd take anything at this point :(

**Comment by ledoge** [comment]
I don't think reloading every X seconds without any heuristics is a good idea, as the periodic flashing will be very annoying and distracting. Might play around with trying to detect events like alt-tabbing at some point. In the meantime, someone reported that apparently driver version `511.23` works fine. so please try that and report back whether the issue also exists there or not.

**Comment by DerAmpelmann** [comment]
I've tried driver version 511.23 by now but unfortunately it did not solve the issue.

**Comment by ledoge** [comment]
Very weird... Does any configuration result in it working properly on your machine? E.g. setting a lower resolution/refresh rate/bit depth, using HDMI instead of DisplayPort, disabling G-Sync, anything like that. Would be good if we could at least narrow it down.

**Comment by DerAmpelmann** [comment]
Ahhhh I found out why/when exactly it happens. Apparently it is caused by switching between G-Sync and fixed refresh rate!

When I disable G-Sync globally and use fixed refresh rate instead, the clamp still works perfectly after alt-tabbing and closing all previously affected fullscreen games. When disabling G-Sync individually for the games where it previously worked fine, it suddenly doesn't work anymore (G-Sync is running in regular desktop usage) and the games that previously broke the clamp work fine when setting those to use G-Sync.).

So the clamp breaks when the monitor switches between G-Sync and fixed refresh. I use G-Sync globally and disable it for individual games (which are the games where the clamp breaks after alt-tabbing/closing the game). This also explains why it only affects esports games for me, as those are the games where I don't use G-Sync :)

Anything you can do to fix this?

**Comment by nicko88** [comment]
I have noticed that I lose the clamp (and profile) when my monitor goes to sleep and I wake it back up.  I have it go to sleep often as it's an OLED.

Personally I would request there to just be a way to re-engage the clamp via a command line parameter.

That way I can just set up a shortcut or some other process of my own to reload the clamp / profile when I need to.


Also, I need to turn off the clamp and profile easily as I have noticed that it breaks HDR.  Well, it breaks HDR when I am using HDR passthrough with madVR.  If I leave the clamp / profile active and try to play an HDR video with madVR, the monitor remains ion SDR mode and my media player (MPC-HC with madVR) never engages HDR mode and passes the HDR through to trigger the monitor to swithc to HDR mode.

If I turn off the clamp / profile then it works as normal.


So yes, if there was a command line parameter which I could use to turn on and off the clamp and profile when I need, for a specific monitor that would be excellent!

**Comment by nicko88** [comment]
A follow-up to my previous post.

If I have G-Sync enabled for fullscreen+windowed and my monitor (Dell AW3423DW) goes to sleep (via Windows power settings) and I wake it up by moving the mouse, the clamp and profile are gone (even though if I open novideo_srgb and look, the checkbox is still checked).  So I have to toggle it off and back on to regain the clamp and profile.

If I turn off G-Sync, then my monitor can go to sleep and wake and I never lose the clamp and profile.

I don't know if the helps.

**Comment by DerAmpelmann** [comment]
Any news on this? Is it fixable within the application or do we have to pray for NVIDIA to fix it in their API?

**Comment by ledoge** [comment]
Please test if any of these issues are fixed in the latest release.

**Comment by Ulf3000** [comment]
> I don't think reloading every X seconds without any heuristics is a good idea, as the periodic flashing will be very annoying and distracting. Might play around with trying to detect events like alt-tabbing at some point. In the meantime, someone reported that apparently driver version `511.23` works fine. so please try that and report back whether the issue also exists there or not.

maybe instead you could check every second IF the clamp is overwritten, and then reload

**Comment by ledoge** [comment]
>maybe instead you could check every second IF the clamp is overwritten, and then reload

The problem with these issues is that NVAPI still reports the clamp as being active, so that doesn't work.

**Comment by jxlarrea** [comment]
@ledoge The clamp is lost when the monitor comes back from sleep. Perhaps you can detect this within the app and reapply the clamp?

**Comment by jxlarrea** [comment]
@nicko88 I have the exact same issue with my AW3423DW OLED. When the display goes to sleep, the clamp is lost after waking it up. I created [a small app](https://github.com/jxlarrea/RunOnDisplayWakeup) that takes care of this issue, It checks every 5 seconds if the monitor is not active and when it comes active again, kills the novideo_srgb process and restarts it so it automatically applies the clamp again.

Gets the job done and works 100% of the time but hopefully this behavior will be included in novideo_srgb itself soon.
